# Star Wars Survey

While waiting for the movie to arrive: *Star Wars: The Force Awakens*, the team at [FiveThirtyEight](http://fivethirtyeight.com/) became interested in the survey about *Star Wars* fans. In particular, they wondered: **does the rest of America realize that "The Empire Strikes Back" is clearly the best of the bunch?**.

The team used the online tool SurveyMonkey to collect the data from *Star War* fans. They received 835 total responses, which are available on [their GitHub repository](https://github.com/fivethirtyeight/data/tree/master/star-wars-survey).

## Project and Data Overview

For this project, we'll clean and explore the dataset to analyze the fan's responses we got about the Star Wars franchise and the movie itself.

from Star Wars fans.

The dataset has several columns, including the following:

- `RespondentID` - An anonymized ID for the respondent (person taking the survey)
- `Gender` - the respondent's gender
- `Age` - the respondent's age
- `Household Income` - the respondent's income
- `Education` - the respondent's eduction level
- `Location (Census Region)` - the respondent's location
- `Have you seen any of the 6 films in the Star Wars franchise` - a `Yes` or `No` respone
- `Do you consider yourself to be a fan of the Star Wars film franchise` - a `Yes` or `No` response

There are several other columns containing answers to questions aobut the *Star Wars* movies. For some questions, the respondent had to check one or more boxes. This type of data is difficult to represent in a column. As a result, this dataset needs a lot of cleaning.

## Import Libraries and Load the Dataset

We need to specify an `encoding=ISO-8859-1`, because the dataset has some characters that aren't in Python's default `utf-8` encoding.

Let's import the required libraries and read in the data to explore.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

%matplotlib inline

In [2]:
# Read the data
star_wars = pd.read_csv('StarWars.csv', encoding='ISO-8859-1')

# Display number of total rows and columns
print(f'Total number of rows: {star_wars.shape[0]}')
print(f'Total number of columns: {star_wars.shape[1]}')

# Explore the data by viewing first ten rows
star_wars.head(10)

Total number of rows: 1186
Total number of columns: 38


RespondentID Have you seen any of the 6 films in the Star Wars franchise?  \
0    3292879998                                                Yes             
1    3292879538                                                 No             
2    3292765271                                                Yes             
3    3292763116                                                Yes             
4    3292731220                                                Yes             
5    3292719380                                                Yes             
6    3292684787                                                Yes             
7    3292663732                                                Yes             
8    3292654043                                                Yes             
9    3292640424                                                Yes             

  Do you consider yourself to be a fan of the Star Wars film franchise?  \
0                                                Yes                      
1                                                NaN                      
2                                                 No                      
3                                                Yes                      
4                                                Yes                      
5                                                Yes                      
6                                                Yes                      
7                                                Yes                      
8                                                Yes                      
9                                                 No                      

  Which of the following Star Wars films have you seen? Please select all that apply.  \
0           Star Wars: Episode I  The Phantom Menace                                    
1                                                NaN                                    
2           Star Wars: Episode I  The Phantom Menace                                    
3           Star Wars: Episode I  The Phantom Menace                                    
4           Star Wars: Episode I  The Phantom Menace                                    
5           Star Wars: Episode I  The Phantom Menace                                    
6           Star Wars: Episode I  The Phantom Menace                                    
7           Star Wars: Episode I  The Phantom Menace                                    
8           Star Wars: Episode I  The Phantom Menace                                    
9                                                NaN                                    

                                    Unnamed: 4  \
0  Star Wars: Episode II  Attack of the Clones   
1                                          NaN   
2  Star Wars: Episode II  Attack of the Clones   
3  Star Wars: Episode II  Attack of the Clones   
4  Star Wars: Episode II  Attack of the Clones   
5  Star Wars: Episode II  Attack of the Clones   
6  Star Wars: Episode II  Attack of the Clones   
7  Star Wars: Episode II  Attack of the Clones   
8  Star Wars: Episode II  Attack of the Clones   
9  Star Wars: Episode II  Attack of the Clones   

                                    Unnamed: 5  \
0  Star Wars: Episode III  Revenge of the Sith   
1                                          NaN   
2  Star Wars: Episode III  Revenge of the Sith   
3  Star Wars: Episode III  Revenge of the Sith   
4  Star Wars: Episode III  Revenge of the Sith   
5  Star Wars: Episode III  Revenge of the Sith   
6  Star Wars: Episode III  Revenge of the Sith   
7  Star Wars: Episode III  Revenge of the Sith   
8  Star Wars: Episode III  Revenge of the Sith   
9                                          NaN   

                          Unnamed: 6  \
0  Star Wars: Episode IV  A New Hope   
1                                NaN   
2                                NaN   
3  Star Wars: Episode IV  A New Hope   
4  Star Wars: 

In [3]:
# View the column names
star_wars.columns.tolist()

['RespondentID',
 'Have you seen any of the 6 films in the Star Wars franchise?',
 'Do you consider yourself to be a fan of the Star Wars film franchise?',
 'Which of the following Star Wars films have you seen? Please select all that apply.',
 'Unnamed: 4',
 'Unnamed: 5',
 'Unnamed: 6',
 'Unnamed: 7',
 'Unnamed: 8',
 'Please rank the Star Wars films in order of preference with 1 being your favorite film in the franchise and 6 being your least favorite film.',
 'Unnamed: 10',
 'Unnamed: 11',
 'Unnamed: 12',
 'Unnamed: 13',
 'Unnamed: 14',
 'Please state whether you view the following characters favorably, unfavorably, or are unfamiliar with him/her.',
 'Unnamed: 16',
 'Unnamed: 17',
 'Unnamed: 18',
 'Unnamed: 19',
 'Unnamed: 20',
 'Unnamed: 21',
 'Unnamed: 22',
 'Unnamed: 23',
 'Unnamed: 24',
 'Unnamed: 25',
 'Unnamed: 26',
 'Unnamed: 27',
 'Unnamed: 28',
 'Which character shot first?',
 'Are you familiar with the Expanded Universe?',
 'Do you consider yourself to be a fan of the Expan

The dataset contains 1186 rows and 38 columns. Note many columns are not descriptive (i.e., `Unnamed`). These columns might be the questions in which the respondent had to check one or more boxes. It's only our assumption at the moment, and we'll find out about them as we make progress in the project.

## Data Cleaning

Take a look at the following two columns:

- `Have you seen any of the 6 films in the Star Wars franchise?`
- `Do you consider yourself to be a fan of the Star Wars film franchise?`

Both represent `Yes/No` questions. They can also be `NaN` where a repondant chooses not a answer a question. We can use the [pandas.Series.value_counts()](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.value_counts.html) method on a series to see all of the unique values in a column, along with the total number of times each value appears.

Let's see the values frequency in these columns:

In [4]:
star_wars['Have you seen any of the 6 films in the Star Wars franchise?'].value_counts(dropna=False)

Yes    936
No     250
Name: Have you seen any of the 6 films in the Star Wars franchise?, dtype: int64

In [5]:
star_wars['Do you consider yourself to be a fan of the Star Wars film franchise?'].value_counts(dropna=False)

Yes    552
NaN    350
No     284
Name: Do you consider yourself to be a fan of the Star Wars film franchise?, dtype: int64

Next, let's find out what is the data type of these columns using [pandas.Series.dtype](https://pandas.pydata.org/docs/reference/api/pandas.Series.dtype.html):

In [6]:
star_wars['Have you seen any of the 6 films in the Star Wars franchise?'].dtype

dtype('O')

In [7]:
star_wars['Do you consider yourself to be a fan of the Star Wars film franchise?'].dtype

dtype('O')

Both columns are currently string types, because the main values they contain are `Yes` and `No`. We can make the data a bit easier to analyze later be converting each column to a Boolean with only the values `True`, `False`, and `NaN`. Booleans are easier to work with because we can select the rows that are `True` or `False` without having to do a string comparison.

We can use the [pandas.Series.map()](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.map.html) method on series objects to perform the conversion.

For example, imagine we have a series that looks like this:

```python
series = ["Yes", "No", NaN, "Yes"]
```

We can use a dictionary to define a mapping from each value in `series` to a new value:

```python
yes_no = {
    "Yes": True,
    "No": False
}
```

Then, we can call the map() function to perform the mapping:

```python
series = series.map(yes_no)
```

`series` will look like this:

```python
[True, False, NaN, True]
```

Let's perform these actions on `Have you seen any of the 6 films in the Star Wars franchise?` and `Do you consider yourself to be a fan of the Star Wars film franchise?` columns:

In [8]:
# Create a dictionary of boolean where 'Yes' is 'True' and
# 'No' is 'False'
yes_no = {
    'Yes': True,
    'No': False
}

# Call series 'map()' function to perform the mapping
star_wars['Have you seen any of the 6 films in the Star Wars franchise?'] = star_wars['Have you seen any of the 6 films in the Star Wars franchise?'].map(yes_no)
star_wars['Do you consider yourself to be a fan of the Star Wars film franchise?'] = star_wars['Do you consider yourself to be a fan of the Star Wars film franchise?'].map(yes_no)

In [9]:
# View the values of these columns after performing the mapping
display(star_wars['Have you seen any of the 6 films in the Star Wars franchise?'].value_counts(dropna=False))
display(star_wars['Do you consider yourself to be a fan of the Star Wars film franchise?'].value_counts(dropna=False))

True     936
False    250
Name: Have you seen any of the 6 films in the Star Wars franchise?, dtype: int64

True     552
NaN      350
False    284
Name: Do you consider yourself to be a fan of the Star Wars film franchise?, dtype: int64

The next six columns represent a single checkbox question. The respondent checked off a series of boxes in response to the question, `Which of the following Star Wars films have you seen? Please select all that apply.`

The columns for this question are:

- `Which of the following Star Wars films have you seen? Please select all that apply.` - whether or not the respondent saw `Star Wars: Episode I The Phantom Menace`.
- `Unnamed: 4` - whether or not the respondent saw `Star Wars: Episode II Attack of the Clones`. 
- `Unnamed: 5` - whether or not the respondent saw `Star Wars: Episode III Revenge of the Sith`. 
- `Unnamed: 6` - whether or not the respondent saw `Star Wars: Episode IV A New Hope`. 
- `Unnamed: 7` - whether or not the respondent saw `Star Wars: Episode V The Empire Strikes Back`. 
- `Unnamed: 8` - whether or not the respondent saw `Star Wars: Episode VI Return of the Jedi`.

For each of these columns, if the value in a cell is the name of the movie, that means the respondent saw the movie. If the value is `NaN`, the respondent either didn't answer or didn't see the movie. We'll assume that they didn't see the movie.

We'll need to convert each of these columns to a Boolean, then rename the column something more intuitive. We can convert the values the same way we did earlier, except that we'll need to include the movie title and `NaN` in the mapping dictionary.

For example, imagine we had this column series:

```python
["Star Wars: Episode I  The Phantom Menace", NaN, "Star Wars: Episode I  The Phantom Menace"]
```

We could convert the values using this mapping dictionary:

```python
{
    "Star Wars: Episode I  The Phantom Menace": True,
    np.NaN: False
}
```

After calling the `map()` method on a series, the column should only contain the values `True` and `False`.

We have to be very careful with spacing when constructing our mapping dictionary! In the cells, `Star Wars: Episode I  The Phantom Menace` has two spaces between the end of `Episode I` and the start of `The Phantom`, but this isn't the case in `Star Wars: Episode VI Return of the Jedi`. We'll check the values in the cells carefully to make sure the appropriate spacing is being used.

In [10]:
# Create dictionary to perform mapping on repective values
mapping_dict = {
    'Star Wars: Episode I  The Phantom Menace': True,
    'Star Wars: Episode II  Attack of the Clones': True,
    'Star Wars: Episode III  Revenge of the Sith': True,
    'Star Wars: Episode IV  A New Hope': True,
    'Star Wars: Episode V The Empire Strikes Back': True,
    'Star Wars: Episode VI Return of the Jedi': True,
    np.NaN: False
}

# Convert each column above so that it only contains the values
# 'True' and 'False'
for col in star_wars.columns[3:9]:
    star_wars[col] = star_wars[col].map(mapping_dict)

Next, we'll need to rename the above columns to reflect what they represent. We can use the [pandas.DataFrame.rename()](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.rename.html) method on DataFrames to do this.

We'll rename each of the columns so the names are more intuitive. The intutition we would like to use so `seen_1` to indicate whether the respondent saw `Star Wars: Episode I The Phantom Menace`, `seen_2` for `Star Wars: Episode II Attack of the Clones`, and so on.

In [15]:
# Display columns to rename
cols_to_rename = star_wars.columns[3:9].tolist()
cols_to_rename

['seen_1', 'seen_2', 'seen_3', 'seen_4', 'seen_5', 'seen_6']

In [13]:
# Create a dictionary of columns to rename
cols_dict = {
    'Which of the following Star Wars films have you seen? Please select all that apply.': 'seen_1',
    'Unnamed: 4': 'seen_2',
    'Unnamed: 5': 'seen_3',
    'Unnamed: 6': 'seen_4',
    'Unnamed: 7': 'seen_5',
    'Unnamed: 8': 'seen_6'
}

# Rename the above columns
star_wars.rename(columns=cols_dict, inplace=True)

In [16]:
# Display results 
star_wars[cols_to_rename].head()

seen_1  seen_2  seen_3  seen_4  seen_5  seen_6
0    True    True    True    True    True    True
1   False   False   False   False   False   False
2    True    True    True   False   False   False
3    True    True    True    True    True    True
4    True    True    True    True    True    True